tests/robustness/validate/validate.go

```diff
@@ -52,6 +56,155 @@ type Config struct {
 	ExpectRevisionUnique bool
 }
 
+func checkValidationAssumptions(reports []report.ClientReport) error {
+	err := putOperationUnique(reports)
+	if err != nil {
+		return err
+	}
+	err = emptyDatabaseAtStart(reports)
+	if err != nil {
+		return err
+	}
+	err = lastOperationSucceededAndObservedInWatch(reports)
+	if err != nil {
+		return err
+	}
+	err = observedAllRevisionsInWatch(reports)
+	if err != nil {
+		return err
+	}
+	err = nonConcurrentClientRequests(reports)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func putOperationUnique(reports []report.ClientReport) error {
+	type KV struct {
+		Key   string
+		Value model.ValueOrHash
+	}
+	putValue := map[KV]struct{}{}
+	for _, r := range reports {
+		for _, op := range r.KeyValue {
+			request := op.Input.(model.EtcdRequest)
+			if request.Type != model.Txn {
+				continue
+			}
+			for _, op := range append(request.Txn.OperationsOnSuccess, request.Txn.OperationsOnFailure...) {
+				if op.Type != model.PutOperation {
+					continue
+				}
+				kv := KV{
+					Key:   op.Put.Key,
+					Value: op.Put.Value,
+				}
+				if _, ok := putValue[kv]; ok {
+					return fmt.Errorf("non unique put %v, required to patch operation history", kv)
+				}
+				putValue[kv] = struct{}{}
+			}
+		}
+	}
+	return nil
+}
+
+func emptyDatabaseAtStart(reports []report.ClientReport) error {
+	for _, r := range reports {
+		for _, op := range r.KeyValue {
+			response := op.Output.(model.MaybeEtcdResponse)
+			if response.Revision == 2 {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("non empty database at start or first write didn't succeed, required by model implementation")
+}
+
+func lastOperationSucceededAndObservedInWatch(reports []report.ClientReport) error {
+	var lastOperation porcupine.Operation
+
+	for _, r := range reports {
+		for _, op := range r.KeyValue {
+			if op.Call > lastOperation.Call {
+				lastOperation = op
+			}
+		}
+	}
+	response := lastOperation.Output.(model.MaybeEtcdResponse)
+	if response.PartialResponse || response.Error != "" {
+		return fmt.Errorf("last operation %v failed, required to validate watch", lastOperation)
+	}
+	request := lastOperation.Input.(model.EtcdRequest)
+	if request.Type != model.Txn || len(request.Txn.Conditions) != 0 || len(request.Txn.OperationsOnSuccess) != 1 || request.Txn.OperationsOnSuccess[0].Type != model.PutOperation || request.Txn.OperationsOnSuccess[0].Put.Key != "tombstone" {
```

Comment:
For easy review, maybe we can use the following pattern or use function to check if it's not tombstone request.

```go
if condition_1 ||
   condition_2 ||
   condition_3 {
}
```

Reply:
Done